django/db/models/functions/comparison.py

```diff
@@ -175,21 +179,28 @@ def join(self, args):
         )
 
     def as_postgresql(self, compiler, connection, **extra_context):
-        if not connection.features.is_postgresql_16:
```

Comment:
I think we could just update this line to be 
`if not connection.features.is_postgresql_16 or connection.features.uses_server_side_binding:`

Right now this mostly reverts the essence of the previous commit

Reply:
So the reason I didn't do that is because I didn't want to emit different SQL based on the server side binding setting.

I imagined some scenario where you have a logger inspecting queries, and I thought it would be "surprising" if the emitted SQL changed and some reporting or metric changed just because you flipped a setting in essentially a client application.

Note that it's still using the native json object on postgres 16 even with server side bindings, if you look a little lower in the code.

Reply:
I'd follow Sarah's suggestion and add `or connection.features.uses_server_side_binding`.

Reply:
I believe I have this setup the desired way:

Postgres before 16: uses jsonb and casts keys to text
Postgres 16 and later with server side bindings: uses native json object and casts keys to text
Postgres 16 and later without server side bindings: uses native json object and skips casting keys to text

Reply:
Cross referencing the other thread, minor update.

The desired behavior was actually:

Postgres before 16 AND Postgres 16 with server side bindings: use jsonb_build_object and casts keys to text
Postgres 16 without server side bindings: use native json object and skip casting keys to text

This requires less code to implement and the desire was to minimize the amount of changes required to fix the regression.